File: 747-min-cost-climbing-stairs/min-cost-climbing-stairs.cpp
class Solution {
public:
    int minCostClimbingStairs(vector<int>& cost) {
        int n = cost.size();
        vector<int> vec(n);
        vec[0] = cost[0], vec[1] = cost[1];

        for (int i=2; i<n; i++) {
            vec[i] = min(vec[i-1], vec[i-2]) + cost[i];
        }

        return min(vec[n-1], vec[n-2]);
    }
};

File: 747-min-cost-climbing-stairs/min-cost-climbing-stairs.cs
public class Solution {
    public int MinCostClimbingStairs(int[] cost) {
        var dp = new Dictionary<int, int>();
        return Math.Min(Helper(cost, cost.Length-1, dp), Helper(cost, cost.Length-2, dp));
    }

    int Helper(int[] cost, int index, IDictionary<int, int> dp) {
        if (index < 0) return 0;
        if (dp.ContainsKey(index)) return dp[index];
        return dp[index] = cost[index] + Math.Min(Helper(cost, index-1, dp), Helper(cost, index-2, dp));
    }
}